Name: Aparna Iyer

PRN: 22070126017

Batch: 2022-2026

Branch: AI-ML A1


###**1. Title:**
Linear Classification Using TensorFlow on CIFAR-10 Dataset.


###**2. Objectives:**

1. Understand linear classification and its application with TensorFlow.

2. Implement a linear classifier for the CIFAR-10 dataset using TensorFlow.

###**3. Theory:**

Linear classification involves using a linear decision boundary to classify data points into different categories. For a given input
𝑥
, the classifier predicts the label by applying a weight matrix
𝑊
and bias
𝑏
, i.e.,
𝑦
=
𝑊
𝑥
+
𝑏
. The goal is to minimize the classification error by adjusting these parameters through training.

The CIFAR-10 dataset consists of 60,000 32x32 color images divided into 10 classes. Linear classifiers often perform poorly on complex datasets like CIFAR-10 because of the high variance and complexity of the data, but they serve as a good baseline for comparison with more complex models like neural networks.

In this experiment, TensorFlow is used to implement and train a linear classifier on the CIFAR-10 dataset, demonstrating the principles of supervised learning, optimization, and evaluation in deep learning.

In [ ]:
import tensorflow as tf
from tensorflow.keras import datasets
import numpy as np

# Load CIFAR-10 dataset
(X_train, y_train), (X_test, y_test) = datasets.cifar10.load_data()



170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 4s 0us/step


In [ ]:
# Normalize the data
X_train, X_test = X_train / 255.0, X_test / 255.0

# Flatten the images into vectors (32x32x3 -> 3072)
X_train_flat = X_train.reshape(X_train.shape[0], -1)
X_test_flat = X_test.reshape(X_test.shape[0], -1)


In [ ]:
# Convert labels to one-hot encoding
y_train_one_hot = tf.keras.utils.to_categorical(y_train, 10)
y_test_one_hot = tf.keras.utils.to_categorical(y_test, 10)

# Define the linear classification model (y = Wx + b)
model = tf.keras.models.Sequential([
    tf.keras.layers.Dense(10, input_shape=(3072,), activation='softmax')
])


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
# Compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(X_train_flat, y_train_one_hot, epochs=10, batch_size=32, validation_data=(X_test_flat, y_test_one_hot))


Epoch 1/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.2593 - loss: 2.0397 - val_accuracy: 0.3543 - val_loss: 1.8464
Epoch 2/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.3527 - loss: 1.8494 - val_accuracy: 0.3543 - val_loss: 1.8468
Epoch 3/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 7s 2ms/step - accuracy: 0.3609 - loss: 1.8186 - val_accuracy: 0.3311 - val_loss: 1.8788
Epoch 4/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.3766 - loss: 1.7936 - val_accuracy: 0.3287 - val_loss: 1.9378
Epoch 5/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 8s 2ms/step - accuracy: 0.3794 - loss: 1.7810 - val_accuracy: 0.3696 - val_loss: 1.7797
Epoch 6/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step - accuracy: 0.3824 - loss: 1.7669 - val_accuracy: 0.3554 - val_loss: 1.8343
Epoch 7/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.3875 - loss: 1.7563 - val_accuracy: 0.3132 - val_loss: 1.9629
Epoch 8/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.3918 - loss: 1.7586 - 

In [ ]:
# Evaluate the model

test_loss, test_acc = model.evaluate(X_test_flat, y_test_one_hot, verbose=2)
print(f'Test accuracy: {test_acc * 100:.2f}%')

313/313 - 0s - 1ms/step - accuracy: 0.3596 - loss: 1.8293
Test accuracy: 35.96%


###**4. Conclusion:**

The experiment successfully demonstrated the use of linear classification on the CIFAR-10 dataset using TensorFlow.

While the linear classifier provided a basic model for classification, its performance was limited by the complexity of the CIFAR-10 dataset.

The Accuracy of the Linear Classifier on the CIFAR-10 Dataset is only 35.96%.

**Justification for 35.96% Accuracy:**

The low accuracy is due to the complexity of CIFAR-10, which contains non-linear patterns that a linear classifier can't capture effectively.

**Improvements:**

a. Use Convolutional Neural Networks (CNNs) for better feature extraction.

b. Apply data augmentation like rotations or flips.

c. Try advanced optimizers like Adam.

d. Increase the number of training epochs.

e. Switching to more complex models and techniques will greatly improve accuracy.

This highlights the need for more advanced architectures, such as deep neural networks, for image classification tasks.

In [ ]:
!apt-get install texlive texlive-xetex texlive-latex-extra pandoc

Processing triggers for tex-common (6.17) ...
Running updmap-sys. This may take some time... done.
Running mktexlsr /var/lib/texmf ... done.
Building format(s) --all.
	This may take some time... done.


In [ ]:
!pip install pypandoc

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!jupyter nbconvert --to PDF "/content/drive/MyDrive/Colab Notebooks/DL_Lab_Experiment_4_AparnaIyer.ipynb"

[NbConvertApp] Running bibtex 1 time: ['bibtex', 'notebook']
[NbConvertApp] WARNING | bibtex had problems, most likely because there were no citations
[NbConvertApp] PDF successfully created
[NbConvertApp] Writing 48206 bytes to /content/drive/MyDrive/Colab Notebooks/DL_Lab_Experiment_4_AparnaIyer.pdf
